### ISSUE TITLE
Patching Commandline Arguments

### ISSUE THREAD

**Issue by slyd0g** [usiegl00/tamatoa#8]
+------------------------+
| Describe Your Question |
+------------------------+
==========================
Hello,

I was wondering if there was a way we could patch command line arguments in the shellcode during the make process. I'm not very knowledgeable in assembly but would love to help with this if you could point me in the right direction!

**Comment by usiegl00** [comment]
Here is how to implement argument handling step by step:
- Get the total size of the argv
- Mmap a new section with that size
- Copy that much from argv to the mmap section
- Store the resulting pointer to the mmap section and the argc in registers that get passed all the way into the program
- Munmap the section after the execution

I will take a look at implementing this.

**Comment by usiegl00** [comment]
I have pushed an update to implement this.
To make with arguments: `make all -- argv0 argv1 argv2 ...`
Let me know if you have any questions.

**Comment by slyd0g** [comment]
Finally got some time to testing this.

So I have a simple Mach-O that prints the arguments using CommandLine.arguments in Swift.

When I use make with the command line arguments, I get the command line arguments of the current program. Is this expected? Could I somehow get the command line arguments that I passed into make?

![image](https://user-images.githubusercontent.com/22451367/157104543-d781cfa4-3233-4f6a-86ea-bb9e78ca4e13.png)

**Comment by usiegl00** [comment]
Hmm, my guess is that swift is reading the arguments from the process address space.
The following code from [Stack Overflow](https://stackoverflow.com/a/40187656) worked for me:
```swift
import Foundation
print("Hello.")
let argc = CommandLine.argc
let argv = UnsafeMutableRawPointer(CommandLine.unsafeArgv).bindMemory(to: UnsafeMutablePointer<Int8>.self, capacity: Int(CommandLine.argc))
NSLog("ARGC: %i", argc);
NSLog("ARGV[0]: %s", argv[0]);
NSLog("ARGV[1]: %s", argv[1]);
```

**Comment by slyd0g** [comment]
Yep, that works for me as well. But is it possible for tamatoa to take in "arg1" and "arg2" during the make phase?

For example, could we pass in "arg1" and "arg2" and have those be used within the Mach-O being loaded in-memory?

Let me know if my question doesn't make sense, I may not be wording it well!

![image](https://user-images.githubusercontent.com/22451367/157358557-e0432b16-62e9-4703-a899-a5d216a40d87.png)

**Comment by usiegl00** [comment]
Your question does make sense.

I used Ghidra to analyze the swift macho. The entrypoint does not take any arguments. (`entry(void)` vs `_main(int argc, char *argv[]`)
I will have to copy the argc and argv to the pointers that swift retrieves them from.

The following C program will work:
```c
#include <stdio.h>
int main (int argc, char *argv[])
{
  int count;
  for (count=0; count<argc; count++)
    puts (argv[count]);
}
```

You can keep this issue open.

**Comment by slyd0g** [comment]
I've been messing with patching arguments for in-memory loading as well. Got it working for C/ObjC with the method you mentioned above ^ If you figure it out on macOS, I'd love to see how you did it to try and recreate :D

**Comment by usiegl00** [comment]
Thank you for reviewing and helping out with development. This project has been lots of fun and I would enjoy a quick chat to share context on my interests. See email.

**Comment by slyd0g** [comment]
> Thank you for reviewing and helping out with development. This project has been lots of fun and I would enjoy a quick chat to share context on my interests. See email.

I can hardly call my debugging efforts "helping out with development", but you're welcome! Did you send me an e-mail? You can contact me at jbui006@ucr.edu or on Twitter @slyd0g :D

**Comment by usiegl00** [comment]
Well, the email I sent to jbui006@ucr.edu bounced.
So I sent an email to justin.bui.ee@gmail.com.
Sorry, I don't use Twitter. Is there a different email I can contact you at?

**Comment by slyd0g** [comment]
Awesome work @usiegl00! I just saw your blogpost and wanted to test it out. I'm unable to get arguments to work with a Swift Mach-O still. I've also noticed 3 arguments causes the problem to segfault but 2 does not

![image](https://user-images.githubusercontent.com/22451367/162022241-c46cf3d9-a806-437b-9ae0-f3f91c9b4b46.png)

**Comment by usiegl00** [comment]
Try `make swift`, I am looking to improve the build system in the future.
I will take a look at the argument patcher to see why 3 arguments cause it to crash.

**Comment by slyd0g** [comment]
`make swift` works with 1 argument but doesn't appear to work with more than 1 
![image](https://user-images.githubusercontent.com/22451367/163047304-03b97b02-8d81-46ca-8d16-fb7a6bb88888.png)
